compiler/ir/backend.jvm/lower/src/org/jetbrains/kotlin/backend/jvm/lower/JvmInlineClassLowering.kt

```diff
@@ -679,35 +679,64 @@ private class JvmInlineClassLowering(context: JvmBackendContext) : JvmValueClass
         )
 
         val function = context.inlineClassReplacements.getSpecializedEqualsMethod(irClass, context.irBuiltIns)
-        val left = function.valueParameters[0]
-        val right = function.valueParameters[1]
 
-        function.body = context.createIrBuilder(irClass.symbol).run {
-            val branches = noinlineSubclasses.map {
-                irBranch(
-                    irCallOp(
-                        boolAnd, context.irBuiltIns.booleanType,
-                        irIs(irGet(left), it.owner.defaultType),
-                        irIs(irGet(right), it.owner.defaultType),
-                    ),
-                    irCallOp(equals, context.irBuiltIns.booleanType, irGet(left), irGet(right))
+        with(context.createIrBuilder(function.symbol)) {
+            function.body = irBlockBody {
+                val left = irTemporary(
```

Comment:
The specialized equals method takes two unboxed arguments, so `coerceInlineClasses` here doesn't do anything except inform `ExpressionCodegen` of the new type before storing a copy of the argument in a local. You can probably just coerce the value parameters at each use instead.